weave-js/src/components/PagePanelComponents/Home/Browse3/datasets/DatasetVersionPage.tsx

```diff
@@ -53,12 +90,203 @@ export const DatasetVersionPage: React.FC<{
       typeof viewerData !== 'object' ||
       viewerData === null ||
       Array.isArray(viewerData);
-    if (dataIsPrimitive) {
-      return {_result: viewerData};
-    }
-    return viewerData;
+    return dataIsPrimitive ? {_result: viewerData} : viewerData;
   }, [viewerData]);
 
+  const originalTableDigest = viewerDataAsObject?.rows?.split('/').pop() ?? '';
+
+  const handleEditClick = useCallback(() => setIsEditing(true), []);
+  const handleCancelClick = useCallback(() => setIsEditing(false), []);
+  const handlePublishClose = () => setPublishAnchorEl(null);
+
+  const handlePublishClick = useCallback(
+    (event: React.MouseEvent<HTMLElement>) => {
+      if (
+        editedCellsMap.size > 0 ||
+        deletedRows.length > 0 ||
+        addedRows.size > 0
+      ) {
+        setPublishAnchorEl(event.currentTarget);
+      }
+    },
+    [editedCellsMap, deletedRows, addedRows]
+  );
+
+  const cleanRow = (row: any) => {
+    return Object.fromEntries(
+      Object.entries(row).filter(([key]) => !['id', '_index'].includes(key))
+    );
+  };
+
+  const convertEditsToTableUpdateSpec = useCallback(() => {
+    const updates: TableUpdateSpec[] = [];
+
+    editedRows.forEach((editedRow, rowIndex) => {
+      if (rowIndex !== undefined) {
+        updates.push({pop: {index: rowIndex}});
+        updates.push({
+          insert: {
+            index: rowIndex,
+            row: cleanRow(editedRow),
+          },
+        });
+      }
+    });
+
+    deletedRows
+      .sort((a, b) => b - a)
+      .forEach(rowIndex => {
+        updates.push({pop: {index: rowIndex}});
+      });
+
+    Array.from(addedRows.values())
+      .reverse()
+      .forEach(row => {
+        updates.push({
+          insert: {
+            index: 0,
```

Comment:
ok, you are inserting at the front here. this MUST be the last edit else all your indexes will be off. I would comment that critical item